File: src/collections_and_containers/cpp/vector/src/vector.cpp
#include "vector.h"
#include <stdexcept>

template <class T> Vector<T>::Vector() : _capacity(0), n(0), data(nullptr) {}

template <class T> Vector<T>::Vector(const Vector<T> &v) {
  n = v.n;
  _capacity = v._capacity;
  if (_capacity > 0) {
    data = new T[_capacity];
    for (int i = 0; i < n; i++) {
      data[i] = v.data[i];
    }
  } else {
    data = nullptr;
  }
}

template <class T> Vector<T> &Vector<T>::operator=(const Vector<T> &v) {
  if (this != &v) {
    delete[] data;
    n = v.n;
    _capacity = v._capacity;
    if (_capacity > 0) {
      data = new T[_capacity];
      for (int i = 0; i < n; i++) {
        data[i] = v.data[i];
      }
    } else {
      data = nullptr;
    }
  }
  return *this;
}

template <class T> Vector<T>::~Vector() { delete[] data; }

template <class T> typename Vector<T>::iterator Vector<T>::begin() {
  return data;
}

template <class T> typename Vector<T>::iterator Vector<T>::end() {
  return data + n;
}

template <class T>
typename Vector<T>::const_iterator Vector<T>::begin() const {
  return data;
}

template <class T> typename Vector<T>::const_iterator Vector<T>::end() const {
  return data + n;
}

template <class T> T &Vector<T>::front() {
  if (n == 0)
    throw std::out_of_range("Vector is empty");
  return data[0];
}

template <class T> const T &Vector<T>::front() const {
  if (n == 0)
    throw std::out_of_range("Vector is empty");
  return data[0];
}

template <class T> T &Vector<T>::back() {
  if (n == 0)
    throw std::out_of_range("Vector is empty");
  return data[n - 1];
}

template <class T> const T &Vector<T>::back() const {
  if (n == 0)
    throw std::out_of_range("Vector is empty");
  return data[n - 1];
}

template <class T> void Vector<T>::push_back(const T &v) {
  if (n >= _capacity) {
    // Double capacity (or start with 1 if empty)
    reserve(_capacity == 0 ? 1 : _capacity * 2);
  }
  data[n++] = v;
}

template <class T> bool Vector<T>::empty() const { return n == 0; }

template <class T> void Vector<T>::pop_back() {
  if (n == 0)
    throw std::out_of_range("Vector is empty");
  --n;
}

template <class T> void Vector<T>::reserve(unsigned int newCapacity) {
  if (static_cast<int>(newCapacity) <= _capacity) {
    return;
  }

  T *newData = new T[newCapacity];
  for (int i = 0; i < n; i++) {
    newData[i] = data[i];
  }

  delete[] data;
  data = newData;
  _capacity = static_cast<int>(newCapacity);
}

template <class T> int Vector<T>::size() const { return n; }

template <class T> int Vector<T>::capacity() const { return _capacity; }

template <class T> void Vector<T>::resize(unsigned int newSize) {
  if (static_cast<int>(newSize) > _capacity) {
    reserve(newSize);
  }
  n = static_cast<int>(newSize);
}

template <class T> T &Vector<T>::operator[](unsigned int index) {
  if (index >= static_cast<unsigned int>(n))
    throw std::out_of_range("Index out of range");
  return data[index];
}

template <class T> const T &Vector<T>::operator[](unsigned int index) const {
  if (index >= static_cast<unsigned int>(n))
    throw std::out_of_range("Index out of range");
  return data[index];
}

template <class T> T &Vector<T>::at(unsigned int index) {
  if (index >= static_cast<unsigned int>(n))
    throw std::out_of_range("Index out of range");
  return data[index];
}

template <class T> const T &Vector<T>::at(unsigned int index) const {
  if (index >= static_cast<unsigned int>(n))
    throw std::out_of_range("Index out of range");
  return data[index];
}

template <class T> void Vector<T>::clear() {
  n = 0;
}

template class Vector<int>;
template class Vector<float>;
template class Vector<double>;
template class Vector<char>;

File: src/collections_and_containers/python/vector/src/vector.py
"""Dynamic array (vector) implementation."""

from __future__ import annotations

from typing import Any, Iterator


class Vector:
    """A dynamic array data structure similar to std::vector in C++."""

    def __init__(self, capacity: int = 10000) -> None:
        self.n = 0
        self.capacity = capacity
        self.data: list[Any] = [None] * self.capacity

    def __iter__(self) -> Iterator[Any]:
        for i in range(self.n):
            yield self.data[i]

    def __len__(self) -> int:
        return self.n

    def __getitem__(self, i: int) -> Any:
        if i < 0 or i >= self.n:
            raise IndexError("Index out of range")
        return self.data[i]

    def __setitem__(self, i: int, value: Any) -> None:
        self.data[i] = value

    def size(self) -> int:
        """Return the number of elements in the vector."""
        return self.n

    def empty(self) -> bool:
        """Check if the vector is empty."""
        return self.n == 0

    def front(self) -> Any:
        """Return the first element."""
        if self.n == 0:
            raise IndexError("Index out of range")
        return self.data[0]

    def back(self) -> Any:
        """Return the last element."""
        if self.n == 0:
            raise IndexError("Index out of range")
        return self.data[self.n - 1]

    def push_back(self, value: Any) -> None:
        """Add an element to the end of the vector."""
        if self.n == self.capacity:
            self.reserve(self.capacity * 2)
        self.data[self.n] = value
        self.n += 1

    def pop_back(self) -> Any:
        """Remove and return the last element."""
        if self.n == 0 or self.n > self.capacity:
            raise IndexError("Index out of range")
        self.n -= 1
        return self.data[self.n]

    def resize(self, size: int) -> None:
        """Resize the vector to contain 'size' elements."""
        if size > self.capacity:
            self.reserve(size)
        self.n = size

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Vector):
            return False
        if self.n != other.n:
            return False
        for i in range(self.n):
            if self.data[i] != other.data[i]:
                return False
        return True

    def clear(self) -> None:
        """Remove all elements from the vector."""
        self.n = 0
        self.capacity = 1
        self.data = [None] * self.capacity

    def reserve(self, size: int) -> None:
        """Reserve capacity for at least 'size' elements."""
        self.capacity = size
        new_data: list[Any] = [None] * self.capacity
        for i in range(self.n):
            new_data[i] = self.data[i]
        self.data = new_data
